NT support for narrow() on dim=0

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #108428
* __->__ #108362
* #108370
* #108361

Satisfies request here: https://github.com/pytorch/pytorch/issues/105913#issuecomment-1652249934

jbschlosser committed:
NT support for narrow() on dim=0

[ghstack-poisoned]

cpuhrsch commented on test/test_nestedtensor.py:
```diff
@@ -2003,6 +2003,25 @@ def test_reshape(self, device, dtype):
         )
 
+    @dtypes(torch.float, torch.float16, torch.double)
+    def test_narrow(self, device, dtype):
```
Have you tried narrowing some non-contiguous NestedTensors? It's also ok to disable non-contiguous NestedTensors for now for this narrow.

jbschlosser replied:
Added error test case for non-contiguous input

cpuhrsch commented on aten/src/ATen/native/nested/NestedTensorFactories.cpp:
```diff
@@ -193,5 +193,42 @@ std::vector<at::Tensor> NestedTensor_unbind(
+
+  return at::detail::make_tensor<NestedTensorImpl>(
+      c10::TensorImpl::VIEW,
+      buffer.narrow(0, buffer_offset, buffer.numel() - buffer_offset),
```
This is scary if the storage offset of the input Tensor is not 0. I think `get_unsafe_storage_as_tensor()` should still work. But I'm actually not sure you need to use buffer at all here. I think you can just update the metadata and pass the storage right through. We might need a VIEW constructor for that.

jbschlosser replied:
> This is scary if the storage offset of the input Tensor is not 0.

Maybe I'm misunderstanding what you're saying, but my code actually queries the first storage offset of the input NT and uses that.

> We might need a VIEW constructor for that.

I'm using the view constructor here, no? Or do you mean a view constructor that accepts a storage instead of a tensor?

cpuhrsch replied:
Yep a view constructor that accepts the storage could be useful. But this is just a nit and not blocking if the inputs are contiguous.

jbschlosser committed:
Update on "NT support for narrow() on dim=0"


Satisfies request here: https://github.com/pytorch/pytorch/issues/105913#issuecomment-1652249934

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

cpuhrsch commented:
@pytorchbot merge

davidbuterez commented:
@jbschlosser Thanks, this seems very useful! Unfortunately I ran into this error when using `torch.narrow` when training with nested tensors: `RuntimeError: derivative for aten::narrow is not implemented` (PyTorch version `2.5.0.dev20240909`). Is there any chance this could be looked at? I am using it in the first dimension, as intended.
